unchained-commons/src/main/java/unchained/commons/utils/ReflectionUtils.java

```diff
@@ -20,6 +22,310 @@ public static Type resolveCommonAncestor(List<Type> types) {
+                populateLambdaArgs(functionalInterface, targetType, map);
+            }
+            populateSuperTypeArgs(targetType.getGenericInterfaces(), map, functionalInterface != null);
+            Type genericType = targetType.getGenericSuperclass();
+            Class<?> type = targetType.getSuperclass();
+            while (type != null && !Object.class.equals(type)) {
+                if (genericType instanceof ParameterizedType) {
+                    populateTypeArgs((ParameterizedType) genericType, map, false);
+                }
+                populateSuperTypeArgs(type.getGenericInterfaces(), map, false);
+                genericType = type.getGenericSuperclass();
+                type = type.getSuperclass();
+            }
+            type = targetType;
+            while (type.isMemberClass()) {
+                genericType = type.getGenericSuperclass();
+                if (genericType instanceof ParameterizedType) {
+                    populateTypeArgs((ParameterizedType) genericType, map, functionalInterface != null);
+                }
+                type = type.getEnclosingClass();
+            }
+            if (CACHE_ENABLED) {
+                TYPE_VARIABLE_CACHE.put(targetType, new WeakReference<>(map));
+            }
+        }
+        return map;
+    }
+
+    private static void populateSuperTypeArgs(final Type[] types, final Map<TypeVariable<?>, Type> map, boolean depthFirst) {
+        for (Type type : types) {
+            if (type instanceof ParameterizedType) {
+                ParameterizedType parameterizedType = (ParameterizedType) type;
+                if (!depthFirst) {
+                    populateTypeArgs(parameterizedType, map, depthFirst);
+                }
+                Type rawType = parameterizedType.getRawType();
+                if (rawType instanceof Class) {
+                    populateSuperTypeArgs(((Class<?>) rawType).getGenericInterfaces(), map, depthFirst);
+                }
+                if (depthFirst) {
+                    populateTypeArgs(parameterizedType, map, depthFirst);
+                }
+            } else if (type instanceof Class) {
+                populateSuperTypeArgs(((Class<?>) type).getGenericInterfaces(), map, depthFirst);
+            }
+        }
+    }
+
+    private static void populateTypeArgs(ParameterizedType type, Map<TypeVariable<?>, Type> map, boolean depthFirst) {
+        if (type.getRawType() instanceof Class) {
+            TypeVariable<?>[] typeVariables = ((Class<?>) type.getRawType()).getTypeParameters();
+            Type[] typeArguments = type.getActualTypeArguments();
+            if (type.getOwnerType() != null) {
+                Type owner = type.getOwnerType();
+                if (owner instanceof ParameterizedType) {
+                    populateTypeArgs((ParameterizedType) owner, map, depthFirst);
+                }
+            }
+            for (int i = 0; i < typeArguments.length; i++) {
```

Comment:
Since we are going down the path of resolving generic type for a class, some of these (e.g. `GenericArrayType`) are guaranteed to not occur and we don't need to check them.